### PR TITLE
[large-tiles] Cross block reader changes plus Cross block iterator

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
Latest version of CrossBlockReader, plus CrossBlockIterator.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

File: src/dbnode/persist/fs/cross_block_iterator.go
Comment (linasm) [reply src/dbnode/persist/fs/cross_block_iterator.go:54]: Not exactly. I have to invert the condition for the early `return true`, and make sure that the call to `c.current.Next()` gets short-circuited away if `c.idx < 0`, like this:

```go
	if c.idx >= 0 && c.current.Next() {
		return true
	}

...
```

File: src/dbnode/persist/fs/cross_block_iterator.go
Comment (robskillington) [inline src/dbnode/persist/fs/cross_block_iterator.go:45]: nit: I would do the following:
```go
c := &crossBlockIterator{current: pool.Get(), byteReader: bytes.NewReader(nil)}
c.Reset(nil)
return c
```

Therefore not duplicating the -1 or any other reset logic that gets added to this iterator.

File: src/dbnode/persist/fs/cross_block_reader.go
Comment (robskillington) [inline src/dbnode/persist/fs/cross_block_reader.go:39]: nit: I usually like to put these right about the struct definition (i.e. `type minHeap struct`) since then you can add more in future, but is just preference so defer to whatever you want to do here.

File: src/dbnode/persist/fs/cross_block_reader.go
Comment (robskillington) [inline src/dbnode/persist/fs/cross_block_reader.go:44]: Any reason to use indices instead of just references to the readers themselves?

i.e.
```go
dataFileSetReaders   []DataFileSetReader
pendingDataFileSetReaders   []DataFileSetReader
```

Comment (robskillington) [reply src/dbnode/persist/fs/cross_block_reader.go:44]: (not saying we should change it, just curious as to one vs other approach and whether you think there's advantages to one or the other, typically we've done the latter elsewhere so you can directly call methods on the references instead of then indexing into the other slice)

File: src/dbnode/persist/fs/cross_block_iterator.go
Comment (linasm) [reply src/dbnode/persist/fs/cross_block_iterator.go:45]: 👍

File: src/dbnode/persist/fs/cross_block_reader.go
Comment (linasm) [reply src/dbnode/persist/fs/cross_block_reader.go:39]: 👍

File: src/dbnode/persist/fs/read.go
Comment (robskillington) [inline src/dbnode/persist/fs/read.go:389]: Hm, I thought we had put in protection to see if entry.Offset+entry.Size isn't out of bands here?  Looks like we don't have any here, should we add it?

Ah I see we removed it, can we add it back in? (that is `if entry.Offset+entry.Size > int64(len(r.dataMmap.Bytes)) {`)

File: src/dbnode/persist/fs/cross_block_reader.go
Comment (linasm) [reply src/dbnode/persist/fs/cross_block_reader.go:44]: I guess the only reason is that then I would have to put a `DataFileSetReader` into a `minHeapEntry` which I wanted to keep just a clean value object.

Comment (linasm) [reply src/dbnode/persist/fs/cross_block_reader.go:44]: Also, as you noticed, `minHeapEntry.dataFileSetReaderIndex` is used for ordering of the `minHeap` (to break the ties when the ids are equal), so I would need to have two values in order to follow your suggestion. Guess will leave this as is.

File: src/dbnode/persist/fs/read.go
Comment (linasm) [reply src/dbnode/persist/fs/read.go:389]: My bad, looks like this was lost during merge conflict resolving.
Will fix this, and also go through the comments from the last PR on these files, to make sure nothing else is missing.

Comment (linasm) [reply src/dbnode/persist/fs/read.go:394]: Resurrected the bounds check on mmaped bytes, also checked the status of other comments after the merge, didn't notice anything else that would be missing.

**Review by robskillington — APPROVED** [review]
LGTM

**Pull request merged**
